airsonic-main/src/main/java/org/airsonic/player/service/MediaScannerService.java

```diff
@@ -145,68 +173,76 @@ public synchronized void scanLibrary() {
         }
         scanning = true;
 
-        Thread thread = new Thread("MediaLibraryScanner") {
-            @Override
-            public void run() {
-                doScanLibrary();
-                playlistService.importPlaylists();
-                mediaFileDao.checkpoint();
-            }
-        };
-
-        thread.setPriority(Thread.MIN_PRIORITY);
-        thread.start();
+        ForkJoinPool pool = new ForkJoinPool(scannerParallelism, mediaScannerThreadFactory, null, true);
+        
+        CompletableFuture.runAsync(() -> doScanLibrary(), pool)
+                .thenRunAsync(() -> playlistService.importPlaylists(), pool)
+                .thenRunAsync(() -> mediaFileDao.checkpoint(), pool)
+                .thenRun(() -> pool.shutdown())
+                .thenRun(() -> scanning = false);
     }
 
     private void doScanLibrary() {
         LOG.info("Starting to scan media library.");
-        Instant lastScanned = Instant.now().truncatedTo(ChronoUnit.SECONDS);
+        Instant lastScanned = Instant.now();
         LOG.debug("New last scan date is {}", lastScanned);
 
         try {
-
             // Maps from artist name to album count.
-            Map<String, Integer> albumCount = new HashMap<String, Integer>();
+            Map<String, AtomicInteger> albumCount = new ConcurrentHashMap<>();
+            Map<String, Artist> artists = new ConcurrentHashMap<>();
+            Map<MediaFile, Album> albums = new ConcurrentHashMap<>();
             Genres genres = new Genres();
 
-            scanCount = 0;
+            scanCount.set(0);
             statistics.reset();
 
             mediaFileService.setMemoryCacheEnabled(false);
             indexManager.startIndexing();
 
-            mediaFileService.clearMemoryCache();
-
             // Recurse through all files on disk.
-            for (MusicFolder musicFolder : settingsService.getAllMusicFolders()) {
-                MediaFile root = mediaFileService.getMediaFile(musicFolder.getPath(), false);
-                scanFile(root, musicFolder, lastScanned, albumCount, genres, false);
-            }
```

Comment:
This is in serial order
- Get the first MusicFolder
- scan it
- Get the next MusicFolder
- scan it
... rinse, repeat